### PR TITLE
Define new API for pppoe.

### PR DESCRIPTION
**Pull request by danderson**
Depends on unsubmitted x/sys/unix patch.

### CODE REVIEW

File: internal/pppoe/pppoe.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	// to tear it down when we close the Conn.
+	discovery net.PacketConn
+	// addr is the address of the remote PPPoE concentrator. We use it
+	// during session teardown, but mostly it's there to provide if
+	// someone asks for Conn.RemoteAddr().
+	addr Addr
+}
+
+// New creates a PPPoE Conn on the given interface.
+func New(ctx context.Context, ifName string) (ret *Conn, err error) {
```
Comment (bradfitz) [inline internal/pppoe/pppoe.go]: "ret" and "err error" stutter are both gross in godoc.

Presumably you did this for line 47, which is unfortunate that the language requires you to name things to do that, even when that's not what you want in godoc.

In these cases I often make an exported version that's pretty that just calls an internal version that has an ugly signature with stutter names.

File: internal/pppoe/pppoe.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		return nil, err
+	}
+
+	// Connect the session fd. This doesn't do much, other than allow
+	// a few more ioctl()s to be applied later on.
+	if err = connectSessionFd(sessionFd, ifName, concentratorAddr, sessionID); err != nil {
+		return nil, err
+	}
+
+	ret = &Conn{
```
Comment (bradfitz) [inline internal/pppoe/pppoe.go]: I'd just return this directly, with `}, nil` on last line.

You can name the result parameter `_`.

File: internal/pppoe/pppoe.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		addr: Addr{
+			Interface:        ifName,
+			SessionID:        sessionID,
+			ConcentratorAddr: concentratorAddr,
+		},
+	}
+	return ret, nil
+}
+
+// LocalAddr returns nil. PPPoE Conns don't have an interesting local
```
Comment (bradfitz) [inline internal/pppoe/pppoe.go]: Things might crash with that, but we'll see. Might want to say "returns nil for now" to not over-promise that it'll always be nil.

I seem to recall a net/http crash where we assumed one of these Addrs was non-nil.

File: internal/pppoe/pppoe.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// RemoteAddr returns the address of the connected PPPoE concentrator,
+// as an *Addr.
+func (c *Conn) RemoteAddr() net.Addr {
+	return &c.addr
+}
+
+// Close closes the PPPoE session.
+func (c *Conn) Close() error {
```
Comment (bradfitz) [inline internal/pppoe/pppoe.go:100]: robust io.Closers are safe from double closes. (even more robust ones are safe for concurrent double closes, but that's overkill.)

But you might want to have a "closed bool" at least on Conn.

**Review by bradfitz — APPROVED** [review]
LGTM

Maybe uint16 for session ids instead of ints, per chat.

File: internal/pppoe/pppoe.go (hunk 1 above)
Comment (danderson) [reply internal/pppoe/pppoe.go]: Turns out there's not as many error-return sites as I anticipated when I did things this way. I just removed the defers and manually closed the right things in the right places if needed.

File: internal/pppoe/pppoe.go (hunk 2 above)
Comment (danderson) [reply internal/pppoe/pppoe.go]: Done.

File: internal/pppoe/pppoe.go (hunk 3 above)
Comment (danderson) [reply internal/pppoe/pppoe.go]: Rephrased to be noncommittal.

File: internal/pppoe/pppoe.go (hunk 4 above)
Comment (danderson) [reply internal/pppoe/pppoe.go:100]: Done, added a closed bool.

I think I will have to make this Conn thread-safe at some point, as well as figure out a way to broadcast "hey I received a PADT from the peer" to upper layers. But I'll do stuff as needed while I implement the upper layers.

**Pull request merged**